ssl/statem/statem_lib.c

```diff
@@ -1744,24 +1761,37 @@ int ssl_choose_server_version(SSL *s, CLIENTHELLO_MSG *hello, DOWNGRADE *dgrd)
  *
  * Returns 0 on success or an SSL error reason number on failure.
  */
-int ssl_choose_client_version(SSL *s, int version, int checkdgrd, int *al)
+int ssl_choose_client_version(SSL *s, int version, RAW_EXTENSION *extensions,
+                              int *al)
 {
     const version_info *vent;
     const version_info *table;
     int highver = 0;
+    int origv;
 
-    /* TODO(TLS1.3): Remove this before release */
-    if (version == TLS1_3_VERSION_DRAFT)
-        version = TLS1_3_VERSION;
+    origv = s->version;
+    s->version = version;
```

Comment:
Yes, it's not ideal but we don't currently have a mechanism for passing back "results" from calling `tls_parse_extension` other than by storing it in the `SSL` object. Possibly we could introduce such a thing - but it kind of seems like a sledgehammer-to-crack-a-nut. That means the supported_versions extension has to store the version "somewhere" and s->version seems the sensible place. Alternatively we could introduce a new temporary variable in SSL - but that seems equally messy.

Reply:
Agreed on probably equally messy to use a new temporary variable.
I wonder what it would look like to do at the start
```
origv = s->version
s->version = version
tls_parse_extension(...)
newv = s->version
s->version = origv
```
and then we would only re-write ```s->version``` (again) for explicit change, as opposed to having to back out the change if something goes wrong.
(Obviously I didn't actually flesh out what that would be; maybe it's even worse than this.)

Reply:
Of course that would mean we have to then change it *again* if we don't encounter an error. Which strikes me as quite messy, so I'm not really sure if this is better.